README
NAME
    Path::Classy - Augmented Path::Class (fancy stuff)

SYNOPSIS
        use Path::Classy qw( file );

        my $f = file( 'my/file' );

        my $size = $f->size(); # size in bytes
        my $h_size = $f->size({ format => 'h'}); # eg. '12K'

        my $mtime = $f->mtime; # secs since epoch
        my $mtime_days = $f->mtime({ as => 'days' }); # days since epoch
        my $mtime_dt = $f->mtime({ as => 'dt' }); # as a DateTime

        my $uri = $f->uri; # URI->new( 'file://...' );

DESCRIPTION
        This module is pretty much an experience by now.
        It is quite superfluous.

    This module augments the API of Path::Class objects with a few more
    methods. The idea is: since you turned your paths into objects, why not
    take it as far as possible?

    For example, after creating a Path::Class object:

        use Path::Class qw( file );

        my $f = file( 'foo.txt' );

    there are some info you can get from the file attached to the current
    path:

        my $st = $f->stat;

    But if you want to know if it exists already or its size (preferably in
    a nice format like '1G')? With Path::Class, you have to say

        print "exists already\n" if -e $f;
        printf "size: %s\n", -s $f;
                           # $f->stat->size works too

    The filetest operators are quite nice and short. But if you want more
    consistency, you would expect methods to access this info:

        print "exists already\n" if $f->exists;
        printf "size: %s\n", $f->size;

    `Path::Classy' provides some of these missing methods. It is far
    complete, as it is a very recent code. But it may occasionally be
    useful.

    ATTENTION: `Path::Classy' does not export anything by default, while
    `Path::Class' exports 'file' and 'dir'.

METHODS
    So far, only files were contemplated with extra methods. These are
    described below.

    In the description below, `$f' is typically created with statements like

      use Path::Classy qw( file );
      $f = file( 'Path-Classy-0.001_0.tar.gz' );

    size [ equiv: -s ]
          $size = $f->size(); 
          $size = $f->size(\%options);

        Returns the file size. By the default, it returns the raw byte
        count. A 'format' option is accepted which support the following
        choices:

        h   Returns the size in a human readable format (like '12K', '1G').
            This is done by using Number::Bytes::Human.

        ''  Returns the raw byte size.

        Equivalence to a filetest expression:

          $f->size      ~~     -s $f

        Examples:

          # assume $f is a file with 24475 bytes
          $f->size() # 24475
          $f->size({ format => 'h' }) # '24K'
          $f->size({ format => '' }) # 24475

    exists [ equiv: -e ]
          $yn = $f->exists;

        Tells if the file exists or not.

        Equivalence to a filetest expression:

          $f->exists    ~~     -e $f

    mtime [ equiv: -M ]
          $mtime = $f->mtime;
          $mtime = $f->mtime(\%options);

        Returns the modification time of the file. By default, it returns
        the number of seconds from epoch (as `$f->stat->mtime' does). It
        supports options 'from' and 'as'. The option 'from' specifies the
        time from a certain reference and the possible choices are:

        'epoch'
            The time is specified since the epoch. (The epoch was at 00:00
            January 1, 1970 GMT.)

        'start'
            The time is specified since the start time of the script (as
            given by `$^T').

        The default is 'epoch'.

        The option 'as' determines how the time is returned. The supported
        choices are:

        'sec'
            The time in seconds since the reference specified by 'from'.

        'days'
            The time in days since the reference specifed by 'from'.

        'dt'
            The time as a DateTime object. It should be independent of the
            reference given by 'from'.

        The default is 'sec'.

        With this method, the equivalence to a filetest expression is not
        that nice:

          $f->mtime({ from => 'start', as => 'days' })

        is

          -( -M $f )

        because `M $f' is the script start time minus file modification time
        in days. In turn,

          $f->mtime()  ~~ $f->stat->mtime

        Examples:

          $f->mtime(); # file mod time since epoch (in secs)
          $f->mtime({ as => 'days' }); # idem (in days)
          $f->mtime({ as => 'dt' }); # file mod time as a DateTime

          $f->mtime({ from => 'epoch', as => 'days' }); # same as $f->mtime

          $f->mtime({ from => 'start' }); # file mod time since script start time (in secs)
          $f->mtime({ from => 'start', as => 'days' }); # same as -( -M $f )
          $f->mtime({ from => 'start', as => 'dt' }); # as a DateTime

        When using `as => 'dt'', it should not make any difference which
        reference time ('epoch' or 'start') is chosen (minus computation
        errors). Prefer `from => 'epoch'' (which can be omitted) because
        there are less involved computations.

    atime [ equiv: -A ]
          $atime = $f->atime;
          $atime = $f->atime(\%options);

        Returns the access time of the file. Everything that was said about
        `mtime' applies here. By default, it returns the number of seconds
        from epoch (as `$f->stat->atime' does). And it supports the options
        'from' and 'as' with the same choices as 'mtime' does.

    ctime [ equiv: -C ]
          $ctime = $f->ctime;
          $ctime = $f->ctime(\%options);

        Returns the creation (or inode change) time of the file (for
        filesystems which support it). Everything that was said about
        `mtime' applies here. By default, it returns the number of seconds
        from epoch (as `$f->stat->ctime' does). And it supports the options
        'from' and 'as' with the same choices as 'mtime' does.

    uri
          $uri = $f->uri;

        Returns a 'file:/' URI pointing to the actual file.



Changes
Revision history for Perl distribution Path-Classy.

0.002_0 Feb 26, 2008
        - now methods size, exists, mtime, atime, ctime
          work for dirs too
        - TODO: update documentation

0.001_0 Feb 14, 2008
        - first CPAN release
        - file methods: size, exists, mtime, atime, 
          ctime, uri


Makefile.PL

use 5.006;
use ExtUtils::MakeMaker;

my $EUMM_VERSION = eval $ExtUtils::MakeMaker::VERSION;

WriteMakefile(
    NAME              => 'Path::Classy',
    VERSION_FROM      => 'lib/Path/Classy.pm',
    PREREQ_PM         => {
        'strict'               => 0,
        'warnings'             => 0,
        'Exporter'             => 0,

        'Path::Class'          => 0,
        'Path::Class::File'    => 0,
        'Path::Class::Dir'     => 0,
        'Number::Bytes::Human' => 0,
        'Attribute::Memoize'   => 0,
        'DateTime'             => 0,
        'URI::file'            => 0,

        # test_requires
        'Test::More'           => 0,
    },
  ( $] >= 5.005 ? (
    ABSTRACT_FROM     => 'lib/Path/Classy.pm',
    AUTHOR            => 'Adriano R. Ferreira <ferreira@cpan.org>'
  ) : () ),
  ( $EUMM_VERSION >= 6.31 ? (
    LICENSE => 'perl',
#    EXTRA_META => "recommends:\n" .
#                  "    Test::Pod:                     1.18\n" .
#                  "    Test::Pod::Coverage:           1.04\n" ,
  ) :() ),

);


lib/Path/Classy/Entity.pm

package Path::Classy::Entity;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.002_0';
our @ISA;

BEGIN {
  require Path::Class::Entity;
  @ISA = qw(Path::Class::Entity);
}

use Attribute::Memoize; # XXX Memoize::Attrs

sub _size_formatter :Memoize {
  require Number::Bytes::Human;
  return Number::Bytes::Human->new;
}

sub _format_size {
  my ($sz, $opt) = @_;

  my $format = $opt->{format};
  if ( $format eq 'h' ) {
    return _size_formatter->format( $sz );
  }
  elsif ( !$format ) { # raw bytes (equiv to no format)
    return $sz;
  }
  else {
    die "unknown format '$format'";
  }
}

# -s
sub size {
  my $self = shift;
  my $raw_size = -s $self;
  if ( @_ ) {
    return _format_size( $raw_size, @_ );
  }
  else {
    return $raw_size;
  }
}

# -z
sub is_empty { -z shift }

# -e
sub exists { -e shift; }

sub _format_mode {
  my ($mode, $opt) = @_;
  my $format = $opt->{format};
  if ( !$format ) {
    return $mode;
  }
  elsif ( $format eq 'oct' ) {
    return sprintf '%04o', $mode;
  } 
  # TODO: 'ls'
  else {
    die "unknown mode format '$format'";
  }
}

# $f->mode
# $f->mode({ format => '' })
# $f->mode({ format => 'ls' })
# $f->mode({ format => 'oct' })
sub mode {
  my $self = shift;
  my $raw_mode = $self->stat->mode & 07777; # perldoc -f stat
  if ( @_ ) {
    return _format_mode( $raw_mode, @_ );
  } else {
    return $raw_mode;
  }
}

sub _dt_from_epoch {
  require DateTime;
  return DateTime->from_epoch( epoch => shift );
}

use constant SECS_IN_A_DAY => 60*60*24*1.00;

sub _format_from_epoch {
  my $sec = shift;
  my $opt = shift;
  my $as  = $opt->{as} || 'sec';
  if ( $as eq 'sec' ) {
    # already in seconds
    return $sec;
  } 
  elsif ( $as eq 'days' ) {
    return $sec / SECS_IN_A_DAY;
  }
  elsif ( $as eq 'dt' ) {
    # return a DateTime
    return _dt_from_epoch( $sec );
  }
  else {
    die "unknown time type '$as'";
  }
}

sub _format_from_start {
  my $days = shift;
  my $opt = shift;
  my $as = $opt->{as} || 'sec';
  if ( $as eq 'days' ) {
    # already in days
    return $days;
  } 
  elsif ( $as eq 'sec' ) {
    return $days * SECS_IN_A_DAY;
  } 
  elsif ( $as eq 'dt' ) {
    return _dt_from_epoch( $^T + $days * SECS_IN_A_DAY );
  }
  else {
    die "unknown time type '$as'";
  }
}



# $f->mtime
# $f->mtime({ from => 'epoch' }) 
# $f->mtime({ from => 'start', as => 'days' }); # -M $f
# $f->mtime({ as => 'dt' });
sub mtime {
  my $self = shift;
  if ( @_ ) {
    my $opt = shift;
    my $from = $opt->{from} || 'epoch';
    if ( $from eq 'epoch' ) {
      return _format_from_epoch( $self->stat->mtime, $opt );
    }
    elsif ( $from eq 'start' ) {
      return _format_from_start( -( -M $self ), $opt );
    }
    else {
      die "unknown time ref '$from'";
    }
  } else {
    return $self->stat->mtime;
  }
}

# -A
sub atime {
  my $self = shift;
  if ( @_ ) {
    my $opt = shift;
    my $from = $opt->{from} || 'epoch';
    if ( $from eq 'epoch' ) {
      return _format_from_epoch( $self->stat->atime, $opt );
    }
    elsif ( $from eq 'start' ) {
      return _format_from_start( -A $self, $opt );
    }
    else {
      die "unknown time ref '$from'";
    }
  } else {
    return $self->stat->atime;
  }
}

# -C
sub ctime {
  my $self = shift;
  if ( @_ ) {
    my $opt = shift;
    my $from = $opt->{from} || 'epoch';
    if ( $from eq 'epoch' ) {
      return _format_from_epoch( $self->stat->ctime, $opt );
    }
    elsif ( $from eq 'start' ) {
      return _format_from_start( -C $self, $opt );
    }
    else {
      die "unknown time ref '$from'";
    }
  } else {
    return $self->stat->ctime;
  }

}


# (-r, -R) (-w, -W) (-x, -X) (-o, -O)


1;


lib/Path/Classy.pm

package Path::Classy;

use 5.006;
use strict;
use warnings;

our $VERSION = '0.002_0';

our @ISA;
our @EXPORT_OK;
BEGIN {
  require Path::Class;

  @EXPORT_OK = @Path::Class::EXPORT_OK;

  require Exporter;
  @ISA = qw( Exporter );
}

use Path::Classy::File ();
use Path::Classy::Dir ();

sub file { Path::Classy::File->new(@_) }
sub dir  { Path::Classy::Dir ->new(@_) }
sub foreign_file { Path::Classy::File->new_foreign(@_) }
sub foreign_dir  { Path::Classy::Dir ->new_foreign(@_) }

1;

__END__

=head1 NAME

Path::Classy - Augmented Path::Class (fancy stuff)

=head1 SYNOPSIS

    use Path::Classy qw( file );

    my $f = file( 'my/file' );

    my $size = $f->size(); # size in bytes
    my $h_size = $f->size({ format => 'h'}); # eg. '12K'

    my $mtime = $f->mtime; # secs since epoch
    my $mtime_days = $f->mtime({ as => 'days' }); # days since epoch
    my $mtime_dt = $f->mtime({ as => 'dt' }); # as a DateTime

    my $uri = $f->uri; # URI->new( 'file://...' );

=head1 DESCRIPTION

    This module is pretty much an experience by now.
    It is quite superfluous.

This module augments the API of L<Path::Class> objects with
a few more methods. The idea is: since you turned your paths
into objects, why not take it as far as possible?

For example, after creating a L<Path::Class> object:

    use Path::Class qw( file );

    my $f = file( 'foo.txt' );

there are some info you can get from the file attached
to the current path:

    my $st = $f->stat;

But if you want to know if it exists already or
its size (preferably in a nice format like '1G')?
With Path::Class, you have to say

    print "exists already\n" if -e $f;
    printf "size: %s\n", -s $f;
                       # $f->stat->size works too

The filetest operators are quite nice and short.
But if you want more consistency, you would expect
methods to access this info:

    print "exists already\n" if $f->exists;
    printf "size: %s\n", $f->size;

C<Path::Classy> provides some of these
missing methods. It is far complete, as it is
a very recent code. But it may occasionally be useful.

B<ATTENTION:> C<Path::Classy> does not export anything by default,
while C<Path::Class> exports 'file' and 'dir'.

=head1 METHODS

So far, only files were contemplated with extra methods.
These are described below.

In the description below, C<$f> is typically created
with statements like

  use Path::Classy qw( file );
  $f = file( 'Path-Classy-0.001_0.tar.gz' );

=over 4

=item B<size> [ equiv: -s ]

  $size = $f->size(); 
  $size = $f->size(\%options);

Returns the file size. By the default, it returns
the raw byte count. A 'format' option is accepted
which support the following choices:

=over 4

=item h

Returns the size in a human readable format (like
'12K', '1G'). This is done by using L<Number::Bytes::Human>.

=item ''

Returns the raw byte size.

=back

Equivalence to a filetest expression:

  $f->size      ~~     -s $f

Examples:

  # assume $f is a file with 24475 bytes
  $f->size() # 24475
  $f->size({ format => 'h' }) # '24K'
  $f->size({ format => '' }) # 24475

=item B<exists> [ equiv: -e ]

  $yn = $f->exists;

Tells if the file exists or not.

Equivalence to a filetest expression:

  $f->exists    ~~     -e $f

=item B<mtime> [ equiv: -M ]

  $mtime = $f->mtime;
  $mtime = $f->mtime(\%options);

Returns the modification time of the file. 
By default, it returns the number of seconds
from epoch (as C<< $f->stat->mtime >> does).
It supports options 'from' and 'as'.
The option 'from' specifies the time from 
a certain reference and the possible choices
are:

=over 4

=item 'epoch'

The time is specified since the epoch.
(The epoch was at 00:00 January 1, 1970 GMT.)

=item 'start'

The time is specified since the start
time of the script (as given by C<$^T>).

=back

The default is 'epoch'.

The option 'as' determines how the time
is returned. The supported choices are:

=over 4

=item 'sec'

The time in seconds since the reference
specified by 'from'.

=item 'days'

The time in days since the reference
specifed by 'from'.

=item 'dt'

The time as a L<DateTime> object. It
should be independent of the reference
given by 'from'.

=back

The default is 'sec'.

With this method, the equivalence to a 
filetest expression is not that nice:

  $f->mtime({ from => 'start', as => 'days' })

is 

  -( -M $f )

because C<-M $f> is the script start time
minus file modification time in days. In turn,

  $f->mtime()  ~~ $f->stat->mtime

Examples:

  $f->mtime(); # file mod time since epoch (in secs)
  $f->mtime({ as => 'days' }); # idem (in days)
  $f->mtime({ as => 'dt' }); # file mod time as a DateTime

  $f->mtime({ from => 'epoch', as => 'days' }); # same as $f->mtime

  $f->mtime({ from => 'start' }); # file mod time since script start time (in secs)
  $f->mtime({ from => 'start', as => 'days' }); # same as -( -M $f )
  $f->mtime({ from => 'start', as => 'dt' }); # as a DateTime

When using C<< as => 'dt' >>, it should not make any
difference which reference time ('epoch' or 'start')
is chosen (except for computation errors). Prefer
C<< from => 'epoch' >> (which can be omitted) because
there are less computations involved.

=item B<atime> [ equiv: -A ]

  $atime = $f->atime;
  $atime = $f->atime(\%options);

Returns the access time of the file. 
Everything that was said about C<mtime> applies here.
By default, it returns the number of seconds
from epoch (as C<< $f->stat->atime >> does).
And it supports the options 'from' and 'as'
with the same choices as 'mtime' does.

=item B<ctime> [ equiv: -C ]

  $ctime = $f->ctime;
  $ctime = $f->ctime(\%options);

Returns the creation (or inode change) time of the file
(for filesystems which support it). 
Everything that was said about C<mtime> applies here.
By default, it returns the number of seconds
from epoch (as C<< $f->stat->ctime >> does).
And it supports the options 'from' and 'as'
with the same choices as 'mtime' does.

=item B<uri>

  $uri = $f->uri;

Returns a 'file://' URI pointing to the actual
file.

=back

=head1 BUGS

Please report bugs via CPAN RT L<http://rt.cpan.org/NoAuth/Bugs.html?Dist=Path-Classy>
or L<mailto://bugs-Path-Classy@rt.cpan.org>.

=head1 SEE ALSO

  Path::Class
  File::Spec

=head1 AUTHORS

Adriano R. Ferreira, E<lt>ferreira@cpan.orgE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2008 by Adriano R. Ferreira

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

=cut

